step-capstone/src/scripts/Optimization.js

```diff
@@ -0,0 +1,85 @@
+const BREAKFAST_TIMERANGE = [6, 10];
+const LUNCH_TIMERANGE = [11, 15];
+const DINNER_TIMERANGE = [17, 21];
+
+export const getOptimalRoute = function (travelObjects, origin, destination) {
+    let request = {
+        origin: new window.google.maps.LatLng(origin.coordinates.lat, origin.coordinates.lng),
+        destination: new window.google.maps.LatLng(destination.coordinates.lat, destination.coordinates.lng),
+        waypoints: travelObjects.map(travelObject => {
+            return { location: new window.google.maps.LatLng(travelObject.coordinates.lat, travelObject.coordinates.lng), stopover: true }
+        }),
+        travelMode: "DRIVING",
+        optimizeWaypoints: true
+    }
+    var directionsService = new window.google.maps.DirectionsService();
+
+    return new Promise(res => {
+        directionsService.route(request, function (result, status) {
+            if (status === "OK") {
+                let order = result.routes[0].waypoint_order; // array of indexes in optimized order
+                var travelObjectsInOrder = order.map(index => travelObjects[index]); // create new array with travel objects in order
+
+                travelObjectsInOrder = [origin].concat(travelObjectsInOrder);
+                travelObjectsInOrder = travelObjectsInOrder.concat([destination]);
+
+                for (let i = 0; i < result.routes[0].legs.length; i++) {
+                    travelObjectsInOrder[i].toNextLocation = result.routes[0].legs[i];
+                }
+
+                res(travelObjectsInOrder);
+            }
+        })
+    })
+}
+
+export const createSchedule = function (travelObjects, userPref, displayDate) {
```

Comment:
This function is complex enough and isolated enough that it could probably benefit from a unit test. To do this you might need to further separate into a utility, but it should be worth it.

Reply:
added testing!